README
NAME
    WWW::Shorten::SuPr - shorten or lengthen URLs with http://su.pr

VERSION
    version 0.001

SYNOPSIS
        use WWW::Shorten::SuPr;
        my $url = q{http://perl.org};
        my $short_url = makeashorterlink($url);

        my $long_url  = makealongerlink($short_url); # eq $url

        my $shortlinkwithauth = shortlinkwithauth($url,$login,$api);

        # Su.pr provides a update your twitter and facebook wall with a
        # message and shorten the links provided in the message
        my $socialpost = WWW::Shorten::SuPrsocialpost($msg,$login,$api);

        #Schedule the posts (time should be in unix format)
        my $schedule_socialpost = WWW::Shorten::SuPr->schedule_socialpost($msg,$login,$api,$time);

SUBROUTINES
  makeashorterlink
    The function "makeashorterlink" will call the is.gd web site passing it
    your long URL and will return the shortened link.

  makealongerlink
    The function "makealongerlink" does the reverse of c<makeashorterlink>
    if the link already been hashed in Su.Pr

  shortlinkwithauth
    Authenticated requests can be used to create account based unique short
    URLs used for Su.pr analytics. Authenticated items will appear on your
    Su.pr home page. Which requires 3 parameters

    URL
               Url Entry supported by Su.Pr

    Login
               Username of the su.pr

    APIKey
               Api key which can get found in su.pr setting page

  socialpost
    The c<socialpost> is used to post Su.pr converted messages to associated
    services such as Twitter and Facebook. Authentication is required for
    this API. Which requires 3 parameters

    Message
            msg: blah blah blah http://perl.org
            Note this should not exceded 140 characters

    Login
            Username of the su.pr

    APIKey
            Api key which can get found in su.pr setting page

  schedule_socialpost
        Schedule a C<socialpost> with unix time stamp .
        Which requires 3 parameters

    Message
            msg: blah blah blah http://perl.org
            Note this should not exceeded 140 characters

    Login
            Username of the su.pr

    APIKey
            Api key which can get found in su.pr setting page

    Time
            Unix timestamp of the date and time you wish to post. Posts will be submitted within a 15 minute bucket of the scheduled time.

AVAILABILITY
    The project homepage is <http://p3rl.org/WWW::Shorten::IsGd>.

    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/WWW-Shorten-IsGd/>.

    The development version lives at
    <http://github.com/doherty/WWW-Shorten-IsGd> and may be cloned from
    <git://github.com/doherty/WWW-Shorten-IsGd.git>. Instead of sending
    patches, please fork this project using the standard git and github
    infrastructure.

SOURCE
    The development version is on github at
    <http://github.com/doherty/WWW-Shorten-IsGd> and may be cloned from
    <git://github.com/doherty/WWW-Shorten-IsGd.git>

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <https://github.com/doherty/WWW-Shorten-IsGd/issues>.

AUTHOR
    Anwesh <kanishka@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Anwesh.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for WWW-Shorten-SuPr

{{$NEXT}}
    First version, released on an unsuspecting world.



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'configure_requires' => {
    'ExtUtils::MakeMaker' => '6.30',
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'shorten or lengthen URLs with http://su.pr',
  'dist_author' => [
    'Anwesh <kanishka@cpan.org>'
  ],
  'dist_name' => 'WWW-Shorten-IsGd',
  'dist_version' => '0.001',
  'license' => 'perl',
  'module_name' => 'WWW::Shorten::IsGd',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Carp' => '0',
    'Exporter' => '0',
    'JSON::Any' => '0',
    'WWW::Shorten::generic' => '0',
    'perl' => '5.008'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/WWW/Shorten/SuPr.pm
package WWW::Shorten::SuPr;
use 5.006;
use strict;
use warnings;

use base qw( WWW::Shorten::generic Exporter );
our @EXPORT = qw( makeashorterlink makealongerlink);
use Carp;
use JSON::Any;

our $VERSION = '0.01';

sub makeashorterlink {
    my $url  = shift or croak 'No URL is passed to makeashorterlink';
    my $ua   = __PACKAGE__->ua();
    my $resp = $ua->post( 'http://su.pr/api/shorten', [ longUrl => $url ] );
    croak "Request failed - " . $resp->status_line
        unless $resp->is_success;

    my $json = JSON::Any->new();
    my $obj  = $json->jsonToObj( $resp->content );
    return $obj->{results}->{$url}->{shortUrl}
        if $obj->{errorCode} == 0;

    return ( $obj->{errorMessage} );
}

sub makealongerlink {
    my $url  = shift or croak 'No URL is passed';
    my $ua   = __PACKAGE__->ua();
    my $resp = $ua->post(
        'http://su.pr/api/expand',
        [
            shortUrl => $url,
            version  => 1.0
        ]
    );
    croak "Request failed - " . $resp->status_line
        unless $resp->is_success;

    my $json = JSON::Any->new;
    my $obj  = $json->jsonToObj( $resp->content );
    return ( $obj->{results}->{ (split m{/}, $url)[3] }->{longUrl} )
        if ( $obj->{errorCode} == '0' );

    return ( $obj->{errorMessage} );
}

sub shortlinkwithauth {
    my ( $url, $login, $api ) = @_ or croak 'Failed to pass one of the parameters';
    my $ua   = __PACKAGE__->ua();
    my $resp = $ua->post(
        'http://su.pr/api/shorten',
        [
            longUrl => $url,
            login   => $login,
            apiKey  => $api
        ]
    );
    croak "Request failed - " . $resp->status_line
        unless $resp->is_success;

    my $json = JSON::Any->new;
    my $obj  = $json->jsonToObj( $resp->content );
    return ( $obj->{results}->{$url}->{shortUrl} )
        if ( $obj->{errorCode} == '0' );

    return ( $obj->{errorMessage} );
}

sub socialpost {
    my ( $msg, $login, $api ) = @_ or croak 'Failed to pass one of the parameters';
    my $ua = __PACKAGE__->ua();
    my $resp = $ua->get("http://su.pr/api/post?&msg=$msg&login=$login&apiKey=$api");
    croak "Request failed - " . $resp->status_line
        unless $resp->is_success;

    my $json = JSON::Any->new;
    my $obj  = $json->jsonToObj( $resp->content );
    return ( $obj->{results}->{shortMsg} )
        if ( $obj->{errorCode} == '0' );

    return ( $obj->{errorMessage} );
}

sub schedule_socialpost {
    my ( $msg, $login, $api, $time ) = @_ or croak 'Failed to pass one of the parameters';
    my $ua   = __PACKAGE__->ua();
    my $resp = $ua->get("http://su.pr/api/post?&msg=$msg&login=$login&apiKey=$api&timestamp=$time");
    croak "Request failed - " . $resp->status_line
        unless $resp->is_success;

    my $json = JSON::Any->new;
    my $obj  = $json->jsonToObj( $resp->content );
    return ( $obj->{results}->{shortMsg} )
        if ( $obj->{errorCode} == '0' );

    return ( $obj->{errorMessage} );
}

1;

=pod

=encoding utf-8

=head1 NAME

WWW::Shorten::SuPr - shorten or lengthen URLs with http://su.pr

=head1 VERSION

Version 0.01

=head1 SYNOPSIS

    use WWW::Shorten::SuPr;
    my $url = q{http://perl.org};
    my $short_url = makeashorterlink($url);

    my $long_url  = makealongerlink($short_url); # eq $url

    my $shortlinkwithauth = shortlinkwithauth($url,$login,$api);

    # Su.pr provides a update your twitter and facebook wall with a
    # message and shorten the links provided in the message
    my $socialpost = WWW::Shorten::SuPrsocialpost($msg,$login,$api);

    #Schedule the posts (time should be in unix format)
    my $schedule_socialpost = WWW::Shorten::SuPr->schedule_socialpost($msg,$login,$api,$time);

=head1 SUBROUTINES/METHODS

=head2 makeashorterlink

The function C<makeashorterlink> will call the is.gd web site passing it your
long URL and will return the shortened link.

=head2 makealongerlink

The function C<makealongerlink> does the reverse of c<makeashorterlink> if the
link already been hashed in Su.Pr

=head2 shortlinkwithauth

Authenticated requests can be used to create account based unique short URLs
used for Su.pr analytics. Authenticated items will appear on your Su.pr home
page. Which requires 3 parameters


=over 7

=item URL

    Url Entry supported by Su.Pr

=item Login

    Username of the su.pr
    
=item APIKey

    Api key which can get found in su.pr setting page
    
=back	

=head2 socialpost

The c<socialpost> is used to post Su.pr converted messages to associated services such
as Twitter and Facebook. Authentication is required for this API.
Which requires 3 parameters

=over

=item Message

    msg: blah blah blah http://perl.org
    Note this should not exceded 140 characters
    
=item Login

    Username of the su.pr
    
=item APIKey
    
    Api key which can get found in su.pr setting page
    
=back	

=head2 schedule_socialpost
    
    
    Schedule a C<socialpost> with unix time stamp .
    Which requires 3 parameters
    
=over

=item Message

    msg: blah blah blah http://perl.org
    Note this should not exceeded 140 characters
    
=item Login
    
    Username of the su.pr
    
=item APIKey

    Api key which can get found in su.pr setting page
    
=item Time

    Unix timestamp of the date and time you wish to post. Posts will be submitted within a 15 minute bucket of the scheduled time.
    
=back	

=head1 AUTHOR

Anwesh, C<< <kanishka at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-shorten-supr at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Shorten-SuPr>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

you can raise a issue on git hosting L<https://github.com/Anwesh/WWW-Shorten-SuPr/issues>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.
    perldoc WWW::Shorten::SuPr
You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here) L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Shorten-SuPr>

=item * AnnoCPAN: Annotated CPAN documentation L<http://annocpan.org/dist/WWW-Shorten-SuPr>

=item * CPAN Ratings L<http://cpanratings.perl.org/d/WWW-Shorten-SuPr>

=item * Search CPAN L<http://search.cpan.org/dist/WWW-Shorten-SuPr/>

=item * Git L<https://github.com/Anwesh/WWW-Shorten-SuPr>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Anwesh.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
See http://dev.perl.org/licenses/ for more information.

=cut


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'shorten or lengthen URLs with http://su.pr',
  'AUTHOR' => 'Anwesh <kanishka@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Test::More' => '0'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.30',
    'Module::Build' => '0.3601'
  },
  'DISTNAME' => 'WWW-Shorten-IsGd',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'WWW::Shorten::IsGd',
  'PREREQ_PM' => {
    'Carp' => '0',
    'Exporter' => '0',
    'JSON::Any' => '0',
    'WWW::Shorten::generic' => '0'
  },
  'VERSION' => '0.001',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



